### ISSUE TITLE
[@types/styled-components] - Types not inferred correctly when using extending a component

### ISSUE THREAD

**Issue by uriklar** [DefinitelyTyped/DefinitelyTyped#42877]
See this code sandbox describing the issue:
https://codesandbox.io/s/styled-components-extend-typescript-kkwdl

The extending component doesn't show type errors even though I define it's props directly. 
Many thanks for your work!
@Igorbek  @Igmat @Lavoaster @Jessidhia

**Comment by smith-chris** [comment]
TL;DR; answer, if you use other name than `theme` for your prop it works properly
https://codesandbox.io/s/styled-components-extend-typescript-fnihi

More descriptive answer:
Quickly looking at styled-component type defs its because of `{theme?: T}` prop that comes from WithOptionalTheme,
which gets returned from StyledComponentProps (as part of a conditional)
which is the type for first parameter of StyledComponentBase function type
which is a part of StyledComponent type 
which finally is what ThemedStyledFunction function interface returns
(actually ThemedStyledFunctionBase that ThemedStyledFunction extends from)

This is quite hard to follow & you probably asking yourself now, but is it a bug or it function as it should? 
I just started looking at those typedefs and my brain has melted so I will let someone more knowlegable to answer that :)

The way you can 'debug' that a little bit easier than following styled typedefs is by using awesome `ReturnType` type util:
```
type StyledReturnType = ReturnType<typeof styled>;
type StyledComponentReturnType = ReturnType<StyledReturnType>;
```

Hope that's not too confusing, I did it just to try understand those types myself and thought sharing my findings might be helpful to someone ^.^

**Comment by uriklar** [comment]
:-D I had to pick the one name that causes a bug! 
Thanks a lot! To me it doesn't sound like a bug, but very good to know
